### ISSUE TITLE
no auto-cleanup

### ISSUE THREAD

**Issue by jakdept** [jakdept/lw-ObjStor4cPanel#3]
cPanel's remote backup transporter includes nothing for backup retention - it puts the backups on the server, and never prunes them.

This isn't an issue for sFTP/FTP backups as you can set up a find on a cron to do retention.

This is not an issue for S3 backups because S3 allows for remote retention rules and stuff.

This is an issue for Object Storage because it allows for neither of the above items.

Thinking of adding a new action to the binary - prune - with two arguments - type and the actual value - to push the binary to prune stuff.

Usage would be:

* `script` `prune` `Pwd` `count` `15` `bucketName` `username`

retain 15 sets of folders of backups

* `script` `prune` `Pwd` `age` `15d` `bucketName` `username`

remove all files older than 15 days

* `script` `prune` `Pwd` `size` `15g` `bucketName` `username`

remove files, oldest first, until usage is under 15g

* `script` `prune` `Pwd` `auto` `bucketName` `username`

work just like count, but pull the retain count from `/var/cpanel/backup/config`.

All four of these options will ignore pwd, bucketName, and username - and instead will process against all valid remote transport destinations configured in cPanel - using the bucketName, username, pw, and pwd in there.

I'll also configure the rpm to add `/etc/cron.d/lw-ObjStor4cPanel.cleanup.cron` - a configuration file - to call the auto cleanup method from above every day.

**Comment by jakdept** [comment]
I was wrong. cPanel does do pruning:

```
./cpbackup_transporter.1503018106.log:32:[2017-08-17 21:01:46 -0400] warn [cpbackup_transporter] Unable to prune:  The specified path does not exist: backups/ at /usr/local/cpanel/Cpanel/LoggerAdapter.pm line 27.
./cpbackup_transporter.1503018106.log:33:	Cpanel::LoggerAdapter::warn(Cpanel::LoggerAdapter=HASH(0x154c970), "Unable to prune:  The specified path does not exist: backups/") called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 597
./cpbackup_transporter.1503018106.log:34:	Cpanel::Backup::Queue::transport_backup::attempt_to_prune_destination(Cpanel::Backup::Queue::transport_backup=HASH(0x192b948), Cpanel::Transport::Files::Custom=HASH(0x2037838), 7, undef, Cpanel::LoggerAdapter=HASH(0x154c970)) called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 226
./cpbackup_transporter.1503381866.log:29:[2017-08-22 02:04:25 -0400] info [cpbackup_transporter] Starting a "prune" operation on the "LW Object Storage" destination ID "6d3l0wURqiy925VVsFLTCYtq".
./cpbackup_transporter.1503381866.log:31:[2017-08-22 02:04:25 -0400] info [cpbackup_transporter] Performing prune operation, retaining 7 items on:  LW Object Storage
./cpbackup_transporter.1503381866.log:32:[2017-08-22 02:04:26 -0400] warn [cpbackup_transporter] Unable to prune:  The specified path does not exist: backups/ at /usr/local/cpanel/Cpanel/LoggerAdapter.pm line 27.
./cpbackup_transporter.1503381866.log:33:	Cpanel::LoggerAdapter::warn(Cpanel::LoggerAdapter=HASH(0x21f3428), "Unable to prune:  The specified path does not exist: backups/") called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 597
./cpbackup_transporter.1503381866.log:34:	Cpanel::Backup::Queue::transport_backup::attempt_to_prune_destination(Cpanel::Backup::Queue::transport_backup=HASH(0x209c268), Cpanel::Transport::Files::Custom=HASH(0x2cbeda0), 7, undef, Cpanel::LoggerAdapter=HASH(0x21f3428)) called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 226
./cpbackup_transporter.1502776961.log:29:[2017-08-15 02:02:41 -0400] info [cpbackup_transporter] Starting a "prune" operation on the "LW Object Storage" destination ID "6d3l0wURqiy925VVsFLTCYtq".
./cpbackup_transporter.1502776961.log:31:[2017-08-15 02:02:41 -0400] info [cpbackup_transporter] Performing prune operation, retaining 7 items on:  LW Object Storage
./cpbackup_transporter.1502776961.log:32:[2017-08-15 02:02:41 -0400] warn [cpbackup_transporter] Unable to prune:  The specified path does not exist: backups/ at /usr/local/cpanel/Cpanel/LoggerAdapter.pm line 27.
./cpbackup_transporter.1502776961.log:33:	Cpanel::LoggerAdapter::warn(Cpanel::LoggerAdapter=HASH(0x11e6428), "Unable to prune:  The specified path does not exist: backups/") called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 597
./cpbackup_transporter.1502776961.log:34:	Cpanel::Backup::Queue::transport_backup::attempt_to_prune_destination(Cpanel::Backup::Queue::transport_backup=HASH(0x15c6890), Cpanel::Transport::Files::Custom=HASH(0x10811d8), 7, undef, Cpanel::LoggerAdapter=HASH(0x11e6428)) called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 226
./cpbackup_transporter.1503208986.log:29:[2017-08-20 02:03:06 -0400] info [cpbackup_transporter] Starting a "prune" operation on the "LW Object Storage" destination ID "6d3l0wURqiy925VVsFLTCYtq".
./cpbackup_transporter.1503208986.log:31:[2017-08-20 02:03:06 -0400] info [cpbackup_transporter] Performing prune operation, retaining 7 items on:  LW Object Storage
./cpbackup_transporter.1503208986.log:32:[2017-08-20 02:03:06 -0400] warn [cpbackup_transporter] Unable to prune:  The specified path does not exist: backups/ at /usr/local/cpanel/Cpanel/LoggerAdapter.pm line 27.
./cpbackup_transporter.1503208986.log:33:	Cpanel::LoggerAdapter::warn(Cpanel::LoggerAdapter=HASH(0x1ca3358), "Unable to prune:  The specified path does not exist: backups/") called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 597
./cpbackup_transporter.1503208986.log:34:	Cpanel::Backup::Queue::transport_backup::attempt_to_prune_destination(Cpanel::Backup::Queue::transport_backup=HASH(0x1e5b100), Cpanel::Transport::Files::Custom=HASH(0x28f9af8), 7, undef, Cpanel::LoggerAdapter=HASH(0x1ca3358)) called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 226
./cpbackup_transporter.1503036163.log:29:[2017-08-18 02:02:43 -0400] info [cpbackup_transporter] Starting a "prune" operation on the "LW Object Storage" destination ID "6d3l0wURqiy925VVsFLTCYtq".
./cpbackup_transporter.1503036163.log:31:[2017-08-18 02:02:43 -0400] info [cpbackup_transporter] Performing prune operation, retaining 7 items on:  LW Object Storage
./cpbackup_transporter.1503036163.log:32:[2017-08-18 02:02:43 -0400] warn [cpbackup_transporter] Unable to prune:  The specified path does not exist: backups/ at /usr/local/cpanel/Cpanel/LoggerAdapter.pm line 27.
./cpbackup_transporter.1503036163.log:33:	Cpanel::LoggerAdapter::warn(Cpanel::LoggerAdapter=HASH(0x1f153f8), "Unable to prune:  The specified path does not exist: backups/") called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 597
./cpbackup_transporter.1503036163.log:34:	Cpanel::Backup::Queue::transport_backup::attempt_to_prune_destination(Cpanel::Backup::Queue::transport_backup=HASH(0x216ab20), Cpanel::Transport::Files::Custom=HASH(0x29bf168), 7, undef, Cpanel::LoggerAdapter=HASH(0x1f153f8)) called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 226
```

it's just failing. The actual calls are relative for ls it looks like, and mkdir, and rmdir:

```
ls / backups/2017-07-23 jakdept ZGQ8TGJW940VELT68MEW
mkdir / backups/2017-07-23 jakdept ZGQ8TGJW940VELT68MEW
put / /backup/2017-07-23/system_files.tar backups/2017-07-23/system_files.tar jakdept ZGQ8TGJW940VELT68MEW
ls / backups/2017-07-23 jakdept ZGQ8TGJW940VELT68MEW
put / /backup/2017-07-23/accounts/deleteos.tar.gz backups/2017-07-23/deleteos.tar.gz jakdept ZGQ8TGJW940VELT68MEW
ls / backups/2017-07-23 jakdept ZGQ8TGJW940VELT68MEW
put / /backup/2017-07-23/accounts/wordpress.tar.gz backups/2017-07-23/wordpress.tar.gz jakdept ZGQ8TGJW940VELT68MEW
ls / backups jakdept ZGQ8TGJW940VELT68MEW
ls / backups/2017-07-24 jakdept ZGQ8TGJW940VELT68MEW
mkdir / backups/2017-07-24 jakdept ZGQ8TGJW940VELT68MEW
put / /backup/2017-07-24/system_files.tar backups/2017-07-24/system_files.tar jakdept ZGQ8TGJW940VELT68MEW
ls / backups/2017-07-24 jakdept ZGQ8TGJW940VELT68MEW
put / /backup/2017-07-24/accounts/wordpress.tar.gz backups/2017-07-24/wordpress.tar.gz jakdept ZGQ8TGJW940VELT68MEW
ls / backups/2017-07-24 jakdept ZGQ8TGJW940VELT68MEW
put / /backup/2017-07-24/accounts/deleteos.tar.gz backups/2017-07-24/deleteos.tar.gz jakdept ZGQ8TGJW940VELT68MEW
ls / backups jakdept ZGQ8TGJW940VELT68MEW
ls / backups/2017-07-25 jakdept ZGQ8TGJW940VELT68MEW
mkdir / backups/2017-07-25 jakdept ZGQ8TGJW940VELT68MEW
put / /backup/2017-07-25/system_files.tar backups/2017-07-25/system_files.tar jakdept ZGQ8TGJW940VELT68MEW
ls / backups/2017-07-25 jakdept ZGQ8TGJW940VELT68MEW
put / /backup/2017-07-25/accounts/deleteos.tar.gz backups/2017-07-25/deleteos.tar.gz jakdept ZGQ8TGJW940VELT68MEW
```

**Comment by jakdept** [comment]
got prune to happen, but now:

```
[2017-08-24 14:16:09 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-20, from LW Object Storage
[2017-08-24 14:16:09 -0400] warn [cpbackup_transporter] Error pruning backup/objstor.hostbaitor.com/2017-07-20 from LW Object Storage:  Executed /usr/local/cpanel/scripts/shim_script.sh rmdir / backup/objstor.hostbaitor.com/2017-07-20 jakdept ZGQ8TGJW940VELT68MEW : 2017/08/24 14:16:09 incorrect number of arguments for a rmdir - [6]
 at /usr/local/cpanel/Cpanel/LoggerAdapter.pm line 27.
        Cpanel::LoggerAdapter::warn(Cpanel::LoggerAdapter=HASH(0x17faa10), "Error pruning backup/objstor.hostbaitor.com/2017-07-20 from L"...) called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 670
        Cpanel::Backup::Queue::transport_backup::attempt_to_prune_destination(Cpanel::Backup::Queue::transport_backup=HASH(0x1c7fa08), Cpanel::Transport::Files::Custom=HASH(0x237fd50), 5, undef, Cpanel::LoggerAdapter=HASH(0x17faa10)) called at /usr/local/cpanel/Cpanel/Backup/Queue.pm line 226
        Cpanel::Backup::Queue::transport_backup::process_task(Cpanel::Backup::Queue::transport_backup=HASH(0x1c7fa08), cPanel::TaskQueue::Task=HASH(0x2463dd0), Cpanel::LoggerAdapter=HASH(0x17faa10)) called at /usr/local/cpanel/3rdparty/perl/524/lib64/perl5/cpanel_lib/cPanel/TaskQueue.pm line 629
        eval {...} called at /usr/local/cpanel/3rdparty/perl/524/lib64/perl5/cpanel_lib/cPanel/TaskQueue.pm line 632
        cPanel::TaskQueue::__ANON__() called at /usr/local/cpanel/3rdparty/perl/524/lib64/perl5/cpanel_lib/cPanel/StateFile.pm line 237
        eval {...} called at /usr/local/cpanel/3rdparty/perl/524/lib64/perl5/cpanel_lib/cPanel/StateFile.pm line 237
        cPanel::StateFile::Guard::call_unlocked(cPanel::StateFile::Guard=HASH(0x237fbb8), CODE(0x237fb70)) called at /usr/local/cpanel/3rdparty/perl/524/lib64/perl5/cpanel_lib/cPanel/TaskQueue.pm line 637
        cPanel::TaskQueue::process_next_task(cPanel::TaskQueue=HASH(0x2282cc0)) called at /usr/local/cpanel/bin/cpbackup_transporter line 151
        eval {...} called at /usr/local/cpanel/bin/cpbackup_transporter line 149
[2017-08-24 14:16:09 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-21, from LW Object Storage
[2017-08-24 14:16:09 -0400] warn [cpbackup_transporter] Error pruning backup/objstor.hostbaitor.com/2017-07-21 from LW Object Storage:  Executed /usr/local/cpanel/scripts/shim_script.sh rmdir / backup/objstor.hostbaitor.com/2017-07-21 jakdept ZGQ8TGJW940VELT68MEW : 2017/08/24 14:16:09 incorrect number of arguments for a rmdir - [6]
```

readded a line that somehow got removed?

**Comment by jakdept** [comment]
got it working:

```
[2017-08-24 14:27:01 -0400] info [cpbackup_transporter] Starting a "prune" operation on the "LW Object Storage" destination ID "YRdplPBCpT4b1nxtVx4wFpPK".
[2017-08-24 14:27:01 -0400] info [cpbackup_transporter] Base path for destination is backup/objstor.hostbaitor.com/
[2017-08-24 14:27:01 -0400] info [cpbackup_transporter] Performing prune operation, retaining 5 items on:  LW Object Storage
[2017-08-24 14:27:02 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-20, from LW Object Storage
[2017-08-24 14:27:02 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-21, from LW Object Storage
[2017-08-24 14:27:02 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-22, from LW Object Storage
[2017-08-24 14:27:03 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-23, from LW Object Storage
[2017-08-24 14:27:03 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-24, from LW Object Storage
[2017-08-24 14:27:04 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-25, from LW Object Storage
[2017-08-24 14:27:04 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-26, from LW Object Storage
[2017-08-24 14:27:05 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-27, from LW Object Storage
[2017-08-24 14:27:05 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-28, from LW Object Storage
[2017-08-24 14:27:06 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-29, from LW Object Storage
[2017-08-24 14:27:06 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-30, from LW Object Storage
[2017-08-24 14:27:06 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-07-31, from LW Object Storage
[2017-08-24 14:27:07 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-01, from LW Object Storage
[2017-08-24 14:27:08 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-02, from LW Object Storage
[2017-08-24 14:27:09 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-03, from LW Object Storage
[2017-08-24 14:27:10 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-05, from LW Object Storage
[2017-08-24 14:27:11 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-06, from LW Object Storage
[2017-08-24 14:27:12 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-07, from LW Object Storage
[2017-08-24 14:27:13 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-08, from LW Object Storage
[2017-08-24 14:27:14 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-10, from LW Object Storage
[2017-08-24 14:27:15 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-11, from LW Object Storage
[2017-08-24 14:27:16 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-12, from LW Object Storage
[2017-08-24 14:27:17 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-13, from LW Object Storage
[2017-08-24 14:27:18 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-14, from LW Object Storage
[2017-08-24 14:27:19 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-15, from LW Object Storage
[2017-08-24 14:27:19 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-16, from LW Object Storage
[2017-08-24 14:27:20 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-17, from LW Object Storage
[2017-08-24 14:27:21 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-18, from LW Object Storage
[2017-08-24 14:27:22 -0400] info [cpbackup_transporter] Pruning backup directory:  backup/objstor.hostbaitor.com/2017-08-19, from LW Object Storage
[2017-08-24 14:27:23 -0400] info [cpbackup_transporter] cpbackup_transporter - Checking queue for tasks
[2017-08-24 14:27:23 -0400] info [cpbackup_transporter] cpbackup_transporter - Processing next task
[2017-08-24 14:27:23 -0400] info [cpbackup_transporter] cpbackup_transporter - Checking queue for tasks
[2017-08-24 14:27:23 -0400] info [cpbackup_transporter] cpbackup_transporter - Exiting - the queue has been emptied; no more work to do
[2017-08-24 14:27:23 -0400] info [cpbackup_transporter] cPanel Backup Transporter Queue Daemon is being stopped.
```

**Comment by Duplika** [comment]
Just to confirm on this, cPanel does the prunning but it's not working properly with LW's object storage?

**Comment by jakdept** [comment]
Apoligies, I never updated (well really closed) this issue. This was corrected a while ago - from the changelog:

```
[root@objstor ~]# rpm -q lw-ObjStor4cPanel --changelog
* Wed Aug 23 2017 Jack Hayhurst <jhayhurst@liquidweb.com> - version 1.1.0
- corrected package summary/long name.
- corrected lsdir function so that it now lists folders
- got cPanel backup purging working
- added something to make sure that the backup destination is added just once
```

This was even corrected in #4 . Just never closed out this issue.

cPanel now prunes backups according to cPanel backup settings/configuration.
